### PR TITLE
Improve `min_score` implementation.

### PR DESCRIPTION
**Pull request by jpountz**
This changes how `min_score` is implemented both for `function_score` and the
search request parameter to confirm whether the minimum score is met in the
`matches()` phase of a TwoPhaseIterator.

### CODE REVIEW

**Comment by javanna** [comment]
I had a look at this and it looks good to me. That said I may not have enough lucene knowledge to review the lucene bits :) 
I think I understand where this is going but maybe we can highlight what this new implementation of min_score brings compare to the previous collector based one?

**Comment by jpountz** [comment]
This PR doesn't change much for the top-level `min_score` parameter, things are still executed in the same order and performance should be about the same. However, this change is important for `function_score` because if you intersect a `function_score` query with a filter, we will now make sure the filter also matches before verifying the score, while we could sometimes verify the score first before, even if the filter does not match.

Eventually I want to deprecate the top-level `min_score` parameter in favour of `function_score` (which is much more flexible as you can filter any sub query and not only the top-level one), so I worked on this pull request first in order to make `function_score` more efficient.

**Comment by javanna** [comment]
sounds good thanks a lot for the explanation @jpountz

File: core/src/main/java/org/elasticsearch/index/query/functionscore/FunctionScoreQueryBuilder.java [hunk 1]
```diff
@@ -312,10 +313,17 @@ protected Query doToQuery(QueryShardContext context) throws IOException {
...
-            return new FunctionScoreQuery(query, function, minScore, combineFunction, maxBoost);
+            query = new FunctionScoreQuery(query, function, combineFunction, maxBoost);
+        } else {
+            // in all other cases we create a FiltersFunctionScoreQuery
+            query = new FiltersFunctionScoreQuery(query, scoreMode, filterFunctions, maxBoost, boostMode == null ? DEFAULT_BOOST_MODE : boostMode);
         }
-        // in all other cases we create a FiltersFunctionScoreQuery
-        return new FiltersFunctionScoreQuery(query, scoreMode, filterFunctions, maxBoost, minScore, boostMode == null ? DEFAULT_BOOST_MODE : boostMode);
+
+        if (minScore != null) {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/index/query/functionscore/FunctionScoreQueryBuilder.java]: This is much nicer.

File: core/src/test/java/org/elasticsearch/common/lucene/search/MinScoreQueryTests.java [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+            Term term = randomTerm();
+            Query query = new TermQuery(term);
+            float minScore = randomMinScore(query);
+            Query minScoreQuery = new MinScoreQuery(query, minScore);
+            assertMinScoreEquivalence(query, minScoreQuery, minScore);
+        }
+    }
+
+    public void testFilteredApproxQuery() throws Exception {
+        // same but with a query that exposes approximations
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/common/lucene/search/MinScoreQueryTests.java]: same *as testBasics*

**Comment by nik9000** [comment]
Looks good to me. I had to go read the two phase stuff because it was new to me but this seems right.

Should this go back to lucene?

**Comment by jpountz** [comment]
> Should this go back to lucene?

I'm not confident enough that these function_score APIs are good to open an issue against Lucene, so I like to keep on incubating this functionnality in Elasticsearch for now. Lucene already has some queries that allow to modify scores like BoostedQuery or BoostingQuery and I tend to find them a bit hack-ish...

**Comment by nik9000** [comment]
Makes sense to me.

**Comment by jpountz** [comment]
I found some issues with how this change interacts with caching and normalization when backporting, so I reverted it. I think we still need to implement something similar so that scores are checked in a TwoPhaseIterator but I guess this can only be done at the (Bulk)Scorer level, not Query/Weight.

**Pull request merged**